common/argparse.h

```diff
@@ -0,0 +1,895 @@
+      -> llvm::raw_ostream& {
+    switch (kind) {
+      case ArgKind::Flag:
+        return output << "Boolean";
+      case ArgKind::Integer:
+        return output << "Integer";
+      case ArgKind::String:
+        return output << "String";
+      case ArgKind::OneOf:
+        return output << "OneOf";
+      case ArgKind::MetaActionOnly:
+        return output << "MetaActionOnly";
+      case ArgKind::Invalid:
+        return output << "Invalid";
+    }
+    CARBON_FATAL() << "Corrupt argument kind!";
+  }
+
+  friend auto operator<<(llvm::raw_ostream& output, CommandKind kind)
+      -> llvm::raw_ostream& {
+    switch (kind) {
+      case CommandKind::Invalid:
+        return output << "Invalid";
+      case CommandKind::RequiresSubcommand:
+        return output << "RequiresSubcommand";
+      case CommandKind::Action:
+        return output << "Action";
+      case CommandKind::MetaAction:
+        return output << "MetaAction";
+    }
+    CARBON_FATAL() << "Corrupt command kind!";
+  }
+};
+
+// The internal representation of a parsable argument description.
+struct Args::Arg {
+  using Kind = ArgKind;
+  using ValueActionT =
+      std::function<bool(const Arg& arg, llvm::StringRef value_string)>;
+  using DefaultActionT = std::function<void(const Arg& arg)>;
+
+  explicit Arg(const ArgInfo& info);
+  ~Arg();
+
+  ArgInfo info;
+  Kind kind = Kind::Invalid;
+  bool has_default = false;
+  bool is_required = false;
+  bool is_append = false;
+  bool is_help_hidden = false;
+
+  // Meta action storage, only populated if this argument causes a meta action.
+  ActionT meta_action = {};
+
+  // Storage options depending on the kind.
+  union {
+    // Singular argument storage pointers.
+    bool* flag_storage;
+    ssize_t* integer_storage;
+    llvm::StringRef* string_storage;
+
+    // Appending argument storage pointers.
+    llvm::SmallVectorImpl<ssize_t>* integer_sequence;
+    llvm::SmallVectorImpl<llvm::StringRef>* string_sequence;
+
+    // One-of information.
+    struct {
+      llvm::OwningArrayRef<llvm::StringRef> value_strings;
+      ValueActionT value_action;
+    };
+  };
+
+  // Default values depending on the kind.
+  union {
+    bool default_flag;
+    ssize_t default_integer;
+    llvm::StringRef default_string;
+    struct {
+      DefaultActionT default_action;
+      int default_value_index;
+    };
+  };
+};
+
+// The internal representation of a parsable command description, including its
+// options, positional arguments, and subcommands.
+struct Args::Command {
+  using Kind = CommandBuilder::Kind;
+
+  explicit Command(const CommandInfo& info, Command* parent = nullptr);
+
+  CommandInfo info;
+  Command* parent;
+  ActionT action = {};
+  Kind kind = Kind::Invalid;
+
+  bool is_help_hidden = false;
+
+  llvm::SmallVector<std::unique_ptr<Arg>, 8> options = {};
+  llvm::SmallVector<std::unique_ptr<Arg>, 2> positional_args = {};
+  llvm::SmallVector<std::unique_ptr<Command>, 2> subcommands = {};
```

Comment:
Are the explicit default initializers here doing much?

Can you add some comments explaining how you chose the sizes (2 and 8), versus using the default?

Reply:
(all of this is done now)